include/grpc/event_engine/slice.h

```diff
@@ -0,0 +1,120 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_EVENT_ENGINE_SLICE_H
+#define GRPC_EVENT_ENGINE_SLICE_H
+
+#include <grpc/impl/codegen/port_platform.h>
+
+#include <string>
+#include <utility>
+
+#include <grpc/slice.h>
+
+namespace grpc_event_engine {
+namespace experimental {
+
+class SliceBuffer;
+
+/// A wrapper around \a grpc_slice.
+///
+/// A slice represents a contiguous reference counted array of bytes.
+/// It is cheap to take references to a slice, and it is cheap to create a
+/// slice pointing to a subset of another slice.
+class Slice final {
+  friend class SliceBuffer;
+
+ public:
+  /// Construct an empty slice.
+  Slice() : slice_(grpc_empty_slice()) {}
+  /// Destructor - drops one reference.
+  ~Slice() { grpc_slice_unref(slice_); }
+
+  enum AddRef { ADD_REF };
+  /// Construct a slice from \a slice, adding a reference.
+  Slice(grpc_slice slice, AddRef) : slice_(grpc_slice_ref(slice)) {}
+
+  enum StealRef { STEAL_REF };
+  /// Construct a slice from \a slice, stealing a reference.
+  Slice(grpc_slice slice, StealRef) : slice_(slice) {}
+
+  /// Allocate a slice of specified size
+  explicit Slice(size_t len) : slice_(grpc_slice_malloc(len)) {}
+
+  /// Construct a slice from a copied buffer
+  Slice(const void* buf, size_t len)
+      : slice_(grpc_slice_from_copied_buffer(reinterpret_cast<const char*>(buf),
+                                             len)) {}
+
+  /// Construct a slice from a copied string
+  /* NOLINTNEXTLINE(google-explicit-constructor) */
+  Slice(const std::string& str)
+      : slice_(grpc_slice_from_copied_buffer(str.c_str(), str.length())) {}
+
+  enum StaticSlice { STATIC_SLICE };
+
+  /// Construct a slice from a static buffer
+  Slice(const void* buf, size_t len, StaticSlice)
+      : slice_(grpc_slice_from_static_buffer(reinterpret_cast<const char*>(buf),
+                                             len)) {}
+
+  /// Copy constructor, adds a reference.
+  Slice(const Slice& other) : slice_(grpc_slice_ref(other.slice_)) {}
```

Comment:
I think I prefer:

```
Slice(const Slice&) = delete;
Slice& operator=(const Slice&) = delete;
Slice(Slice&& other) : slice_(other.slice_) { other.slice_ = kEmpty; }
Slice& operator=(Slice&& other) {
  std::swap(slice_, other.slice_);
  return *this;
}
Slice Ref() { return Slice(grpc_slice_ref(slice_)); }
Slice Copy() { return Slice(grpc_slice_copy(slice_)); }
```

i.e. slice itself is move only, and has explicit factory functions for reffing and copying.

this clarifies intent at copy sites, and when substituting this type into existing code we don't accidentally end up with an explosion of refcounting.